scrapy/downloadermiddlewares/cookies.py

```diff
@@ -1,45 +1,69 @@
+from __future__ import annotations
+
 import logging
 from collections import defaultdict
+from http.cookiejar import Cookie
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    DefaultDict,
+    Dict,
+    Iterable,
+    Optional,
+    Sequence,
+    Union,
+)
 
 from tldextract import TLDExtract
 
+from scrapy import Request, Spider
+from scrapy.crawler import Crawler
 from scrapy.exceptions import NotConfigured
 from scrapy.http import Response
 from scrapy.http.cookies import CookieJar
 from scrapy.utils.httpobj import urlparse_cached
 from scrapy.utils.python import to_unicode
 
+if TYPE_CHECKING:
+    # typing.Self requires Python 3.11
+    from typing_extensions import Self
+
+
 logger = logging.getLogger(__name__)
 
 
 _split_domain = TLDExtract(include_psl_private_domains=True)
 
 
-def _is_public_domain(domain):
+def _is_public_domain(domain: str) -> bool:
     parts = _split_domain(domain)
     return not parts.domain
 
 
 class CookiesMiddleware:
     """This middleware enables working with sites that need cookies"""
 
-    def __init__(self, debug=False):
-        self.jars = defaultdict(CookieJar)
-        self.debug = debug
+    def __init__(self, debug: bool = False):
+        self.jars: DefaultDict[Any, CookieJar] = defaultdict(CookieJar)
+        self.debug: bool = debug
 
     @classmethod
-    def from_crawler(cls, crawler):
+    def from_crawler(cls, crawler: Crawler) -> Self:
         if not crawler.settings.getbool("COOKIES_ENABLED"):
             raise NotConfigured
         return cls(crawler.settings.getbool("COOKIES_DEBUG"))
 
-    def _process_cookies(self, cookies, *, jar, request):
+    def _process_cookies(
+        self, cookies: Iterable[Cookie], *, jar: CookieJar, request: Request
+    ) -> None:
         for cookie in cookies:
             cookie_domain = cookie.domain
             if cookie_domain.startswith("."):
                 cookie_domain = cookie_domain[1:]
 
-            request_domain = urlparse_cached(request).hostname.lower()
+            hostname = urlparse_cached(request).hostname
+            assert hostname is not None
```

Comment:
Is this a given? I was thinking some URL protocols do allow this in valid URLs, and I am not sure we prevent cookies from being used with them:

```python
>>> from urllib.parse import urlparse
>>> print(urlparse('file:///home/adrian').hostname)
None
```

Reply:
The code will fail anyway as it calls `.lower()` on it later. But I'm not sure if it's possible to run this code in a non-HTTP context.

Reply:
OK. I guess the only change here is that the exception will be different, and if anyone is getting this exception they should report it anyway so that we can do something about it (maybe prevent cookie handling from working on file:// altogether).